### PR TITLE
sensor: Implement driver and tests for akm09918c

### PR DESCRIPTION
**Pull request by yperess**
Add the implementation for the akm09918c magnetometer driver. Additionally, add the appropriate node to the TDK robokit1 device tree. In order to prevent regressions, add the sensor to the sensor build_all tests and specific tests using an emulator.

### CODE REVIEW

File: drivers/sensor/akm09918c/akm09918c.c [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+	return 0;
+}
+
+static void akm09918c_convert(struct sensor_value *val, int16_t sample)
+{
+	int64_t conv_val = sample * AKM09918C_MICRO_GAUSS_PER_BIT;
+
+	val->val1 = conv_val / 1000000;
+	val->val2 = conv_val % 1000000;
```
Comment (teburd) [inline drivers/sensor/akm09918c/akm09918c.c:69]: Do you really need a modulo here? Multiply and subtract, conv_val - val1*1000000 and you have the remainder.

File: tests/drivers/build_all/sensor/spi.dtsi [hunk 2]
```diff
@@ -337,3 +337,8 @@ test_spi_icm426888: icm42688@28 {
 	gyro-hz = <32000>;
 	gyro-fs = <2000>;
 };
+
+test_spi_akm09918c: akm09918c@29 {
```
Comment (teburd) [inline tests/drivers/build_all/sensor/spi.dtsi:341]: i2c device correct? seems like the wrong file and dts node here

**Review by teburd — CHANGES_REQUESTED** [review]
I guess I'm interested to know what the modulo compiles to, I'd guess a mult and add would be only a few ops... which I'd guess would be fairer to compare to a fixed point thing.

The test dts seems for build_all seems wrong, i2c device not spi

File: drivers/sensor/akm09918c/akm09918c.c (hunk 1 above)
Comment (yperess) [reply drivers/sensor/akm09918c/akm09918c.c:69]: Done

File: tests/drivers/build_all/sensor/spi.dtsi (hunk 2 above)
Comment (yperess) [reply tests/drivers/build_all/sensor/spi.dtsi:341]: Oops :)

File: drivers/sensor/akm09918c/akm09918c.c [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+	if (data->mode == AKM09918C_CNTL2_PWR_DOWN) {
+		if (i2c_reg_write_byte_dt(&cfg->i2c, AKM09918C_REG_CNTL2,
+					  AKM09918C_CNTL2_SINGLE_MEASURE) != 0) {
+			LOG_ERR("Failed to start measurement.");
+			return -EIO;
+		}
+
+		/* Wait for sample */
+		LOG_DBG("Waiting for sample...");
+		k_busy_wait(AKM09918C_MEASURE_TIME_US);
```
Comment (teburd) [inline drivers/sensor/akm09918c/akm09918c.c:43]: this is a relatively long busy wait

**Review by teburd — DISMISSED** [review]
Looks good to me. The long k_busy_wait *maybe* could do with a k_msleep as other tasks could absolutely run in 9ms.

File: drivers/sensor/akm09918c/akm09918c.c (hunk 3 above)
Comment (yperess) [reply drivers/sensor/akm09918c/akm09918c.c:43]: I also thought it was odd but used the other `ak` sensor as reference. I'm happy to change it but assumed there was a hard reason for it.

Comment (MaureenHelm) [reply drivers/sensor/akm09918c/akm09918c.c:43]: Yeah, let's convert that to a sleep

File: drivers/sensor/akm09918c/akm09918c_emul.c [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+static int akm09918c_emul_init(const struct emul *target, const struct device *parent)
+{
+	ARG_UNUSED(parent);
+	akm09918c_emul_reset(target);
+
+	return 0;
+}
+
+static struct i2c_emul_api akm09918c_emul_api_i2c = {
```
Comment (MaureenHelm) [inline drivers/sensor/akm09918c/akm09918c_emul.c:133]: `const`?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/akm09918c/akm09918c.c (hunk 3 above)
Comment (yperess) [reply drivers/sensor/akm09918c/akm09918c.c:43]: Done

File: drivers/sensor/akm09918c/akm09918c_emul.c (hunk 4 above)
Comment (yperess) [reply drivers/sensor/akm09918c/akm09918c_emul.c:133]: Done

File: drivers/sensor/ak8975/ak8975.c [hunk 5]
```diff
@@ -55,7 +55,7 @@ static void ak8975_convert(struct sensor_value *val, int16_t sample,
 	conv_val = sample * AK8975_MICRO_GAUSS_PER_BIT *
 		   ((uint16_t)adjustment + 128) / 256;
 	val->val1 = conv_val / 1000000;
-	val->val2 = conv_val % 1000000;
```
Comment (teburd) [inline drivers/sensor/ak8975/ak8975.c:58]: Seems like this was changed in a different file than the akm09918c file which is being added?

**Review by teburd — CHANGES_REQUESTED** [review]

File: drivers/sensor/ak8975/ak8975.c (hunk 5 above)
Comment (yperess) [reply drivers/sensor/ak8975/ak8975.c:58]: Oops :)

**Review by teburd — DISMISSED** [review]

File: tests/drivers/build_all/sensor/i2c.dtsi [hunk 6]
```diff
@@ -652,3 +652,8 @@ test_i2c_bh1750: bh1750@65 {
 	compatible = "rohm,bh1750";
 	reg = <0x65>;
 };
+
+test_i2c_akm09918c: akm09918c@55 {
```
Comment (MaureenHelm) [inline tests/drivers/build_all/sensor/i2c.dtsi:656]: Address should be 0x66

File: drivers/sensor/akm09918c/akm09918c.c [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+
+	if (buffer[0] != AKM09918C_WIA1 || buffer[1] != AKM09918C_WIA2) {
+		LOG_ERR("Wrong who-am-i value");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+int akm09918c_init(const struct device *dev)
```
Comment (MaureenHelm) [inline drivers/sensor/akm09918c/akm09918c.c:178]: static

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Comment by yperess** [comment]
Latest push was a rebase

**Review by MaureenHelm — APPROVED** [review]

**Review by teburd — APPROVED** [review]

**Comment by teburd** [comment]
Tested and works well on the robokit

**Pull request merged**